skbio/alignment/_tabular_msa.py

```diff
@@ -1095,6 +1095,252 @@ def _assert_valid_sequences(self, sequences):
+        ...                      positional_metadata={'col2': [1, 2, 3, 4, 5]})
+        True
+
+        When performing an outer join (``'outer'``, ``'left'``, or
+        ``'right'``), unshared sequences are padded with gaps and unshared
+        ``positional_metadata`` columns are padded with NaN:
+
+        >>> joined = msa1.join(msa2, how='outer')
+        >>> joined == TabularMSA([DNA('ACT--'),
+        ...                       DNA('A-G-T'),
+        ...                       DNA('-C---'),
+        ...                       DNA('--ACG')], index=['a', 'b', 'c', 'z'],
+        ...                      positional_metadata={
+        ...                          'col1': [42, 43, np.nan, np.nan, np.nan],
+        ...                          'col2': [1, 2, 3, 4, 5],
+        ...                          'col3': [np.nan, np.nan, 'f', 'o', 'o']})
+        True
+
+        """
+        if how not in {'strict', 'inner', 'outer', 'left', 'right'}:
+            raise ValueError(
+                "`how` must be 'strict', 'inner', 'outer', 'left', or "
+                "'right'.")
+
+        self._assert_joinable(other)
+
+        join_index, concat_kwargs = self._get_join_index(other, how)
+
+        joined_seqs = []
+        for label in join_index:
+            left_seq = self._get_sequence_for_join(label)
+            right_seq = other._get_sequence_for_join(label)
+
+            joined_seqs.append(
+                self.dtype.concat([left_seq, right_seq], how='outer'))
+
+        # TODO: update when #1198 is implemented.
+        joined_positional_metadata = None
+        if joined_seqs:
+            joined_positional_metadata = pd.concat(
+                [self.positional_metadata, other.positional_metadata],
+                ignore_index=True, **concat_kwargs)
+
+            if not self.has_positional_metadata():
+                del self.positional_metadata
+            if not other.has_positional_metadata():
+                del other.positional_metadata
+
+        joined = self.__class__(joined_seqs, index=join_index,
+                                positional_metadata=joined_positional_metadata)
+
+        if not joined.has_positional_metadata():
+            del joined.positional_metadata
+
+        return joined
+
+    def _assert_joinable(self, other):
+        if not isinstance(other, TabularMSA):
+            raise TypeError(
+                "`other` must be a `TabularMSA` object, not type %r" %
+                type(other).__name__)
+
+        if self.dtype is not other.dtype:
+            raise TypeError(
+                "`other` dtype %r does not match this MSA's dtype %r" %
+                (other.dtype if other.dtype is None else other.dtype.__name__,
+                 self.dtype if self.dtype is None else self.dtype.__name__))
+
+        if not self.index.is_unique or not other.index.is_unique:
+            raise ValueError(
+                "Cannot join `TabularMSA` objects. Index labels must be "
+                "unique.")
```

Comment:
Same here, users might get the wrong impression as to how unique things have to be.